File: python/petstorePy/petstore.py
#  Copyright NetFoundry Inc.
#
#  Licensed under the Apache License, Version 2.0 (the "License");
#  you may not use this file except in compliance with the License.
#  You may obtain a copy of the License at
#
#  https://www.apache.org/licenses/LICENSE-2.0
#
#  Unless required by applicable law or agreed to in writing, software
#  distributed under the License is distributed on an "AS IS" BASIS,
#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
#  See the License for the specific language governing permissions and
#  limitations under the License.

import getopt
import os
import os.path
import sys
import openziti
import requests

DEFAULT_IDENTITY_FILE = 'taste_of_ziti.json'
APERITIVO_OPT = 'aperitivo'
IDENTITY_OPT = 'identity'
QUERY_OPT = 'query'


def create_aperitivo_identity(opts):
    print('Connecting to aperitivo to generate a new temporary identity')
    if APERITIVO_OPT in opts:
        aperitivo_url = opts[APERITIVO_OPT]
    else:
        aperitivo_url = 'https://aperitivo.production.netfoundry.io'

    aperitivo_response = requests.post(aperitivo_url + '/aperitivo/v1/identities')
    if aperitivo_response.status_code != 200:
        print('Failure to connect to aperitivo to generate a new identity.  Status code = ' +
              str(aperitivo_response.status_code))
        sys.exit(2)

    # Aperitivo returns a json structure like this:
    # {
    #   "name": String identity name,
    #   "jwt": String identity jwt to enroll,
    #   "validUntil": Date string indicating when the demo identity will be automatically removed
    # }
    token_file = open(DEFAULT_IDENTITY_FILE, 'w')
    token_file.write(openziti.enroll(aperitivo_response.json()['jwt']))
    token_file.close()
    print('Saved an openziti identity in ' + DEFAULT_IDENTITY_FILE +
          '. This is a temporary identity that is valid until ' + aperitivo_response.json()['validUntil'])


def check_create_identity(opts):
    # check user input and if specified, return it
    if opts.get(IDENTITY_OPT, '') != '':
        print('Loading identity from ' + opts[IDENTITY_OPT])
        openziti.load(opts[IDENTITY_OPT])

    # If the identity is not specified on the command line, try to pull it from the environment
    elif os.environ.get('ZITI_IDENTITIES', '') != '':
        print('Loading identity from ' + os.environ['ZITI_IDENTITIES'])
        # do not need to call openziti.load here, will be loaded automatically

    # If still empty, create or use the identity from aperitivo
    else:
        if not os.path.exists(DEFAULT_IDENTITY_FILE):
            create_aperitivo_identity(opts)
        print('Loading identity from ' + DEFAULT_IDENTITY_FILE)
        try:
            openziti.load(DEFAULT_IDENTITY_FILE)
        except:
            print('Cannot authenticate with the existing aperitivo identity. Try deleting the previously saved ' +
                  'identity file at \'' + DEFAULT_IDENTITY_FILE + '\' and trying again.')
            sys.exit(1)


def parse_opts(argv):
    parsed_opts = {}
    opts, args = getopt.getopt(argv, "a:hi:q:", ["aperitivo=", "identity=", "query="])
    for opt, arg in opts:
        if opt == '-h':
            print('petstore.py -i <identityFile> -q <petstore query> -a <aperitivoUrl>')
            sys.exit()
        elif opt in ("-a", "--aperitivoUrl"):
            parsed_opts[APERITIVO_OPT] = arg
        elif opt in ("-i", "--identity"):
            parsed_opts[IDENTITY_OPT] = arg
        elif opt in ("-q", "--query"):
            parsed_opts[QUERY_OPT] = arg
    return parsed_opts


def hit_ziti_service(opts):
    # Use openziti monkeypatch to query over ziti
    if QUERY_OPT in opts:
        query_str = opts[QUERY_OPT]
    else:
        query_str = '/'

    with openziti.monkeypatch():
        print('Querying petstore over openziti with query: ' + query_str)
        r = requests.get("http://petstore.ziti" + query_str)
        print(r.text)


if __name__ == "__main__":
    user_opts = parse_opts(sys.argv[1:])

    check_create_identity(user_opts)
    hit_ziti_service(user_opts)

File: golang/petstoreClient/petstore.go
/*
	Copyright NetFoundry Inc.

	Licensed under the Apache License, Version 2.0 (the "License");
	you may not use this file except in compliance with the License.
	You may obtain a copy of the License at

	https://www.apache.org/licenses/LICENSE-2.0

	Unless required by applicable law or agreed to in writing, software
	distributed under the License is distributed on an "AS IS" BASIS,
	WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
	See the License for the specific language governing permissions and
	limitations under the License.
*/

package main

import (
  "context"
  "flag"
  "fmt"
  "github.com/openziti/sdk-golang/ziti"
  "io"
  "log"
  "net"
  "net/http"
  "os"
)

const defaultAperitivoUrl = "https://aperitivo.production.netfoundry.io"

func parseOpts() map[string]string {
  aperitivoPtr := flag.String("a", defaultAperitivoUrl, "optional Aperitivo url for acquiring an identity")
  identityPtr := flag.String("i", "", "optional identity file")
  queryPtr := flag.String("q", "/api/v3/pet/findByStatus?status=available", "petstore query to execute")
  var showHelp bool
  flag.BoolVar(&showHelp, "h", false, "Display usage")
  flag.Parse()
  if showHelp {
    flag.PrintDefaults()
    os.Exit(0)
  }
  parsedOpts := make(map[string]string)
  if *identityPtr != "" {
    parsedOpts["identity"] = *identityPtr
  }
  // Options with default values so just add
  parsedOpts["query"] = *queryPtr
  parsedOpts["aperitivo"] = *aperitivoPtr
  return parsedOpts
}

// Reads the identity JSON file and adds the parsed ziti.Context into a ziti.CtxCollection
func loadIdentity(identityFile string) *ziti.CtxCollection {
  log.Printf("Loading identity from %s", identityFile)
  ctx, err := ziti.NewContextFromFile(identityFile)
  if err != nil {
    log.Fatalf("Unable to load context from %s", identityFile)
  }
  collection := ziti.NewSdkCollection()
  collection.Add(ctx)

  return collection
}

// load or generate and then load the ziti Identity
func checkCreateIdentity(opts map[string]string) *ziti.CtxCollection {
  // First check to see if an identity file was specified on the command line.  If so, use it
  if cmdLineIdentity, present := opts["identity"]; present {
    return loadIdentity(cmdLineIdentity)
  }
  // Next check to see if the default identity file is available. If not, generate an identity into it
  if _, err := os.Stat(DefaultIdentityFile); err != nil {
    GenerateIdentity(opts["aperitivo"])
  }
  // Lastly, load the identity from the generated or pre-existing default identity file
  return loadIdentity(DefaultIdentityFile)
}

func buildZitiClient(collection *ziti.CtxCollection) *http.Client {
  zitiTransport := http.DefaultTransport.(*http.Transport).Clone() // copy default transport
  zitiTransport.DialContext = func(_ context.Context, _, addr string) (net.Conn, error) {
    // Need to search for a context within the collection that provides access to the service name in the address parameter
    service, _, err := net.SplitHostPort(addr)
    if err != nil {
      return nil, err
    }

    var ztx ziti.Context
    var matchFound = false
    collection.ForAll(func(ctx ziti.Context) {
      if matchFound {
        return
      }

      if _, found := ctx.GetService(service); found {
        ztx = ctx
        matchFound = true
      }
    })

    if matchFound {
      return ztx.Dial(service)
    }
    return nil, fmt.Errorf("service [%s] is not available by any ziti context", service)
  }
  zitiTransport.TLSClientConfig.InsecureSkipVerify = true
  return &http.Client{Transport: zitiTransport}
}

func connectZitiService(ctxCollection *ziti.CtxCollection, opts map[string]string) {
  if !CheckIdentityIsStillValid(ctxCollection, "PetstoreDemo") {
    log.Fatal("PetstoreDemo service not accessible by this Identity")
  }
  log.Printf("Calling PetstoreDemo with query string from command line: %s", opts["query"])
  response, err := buildZitiClient(ctxCollection).Get("http://PetstoreDemo" + opts["query"])
  if err != nil {
    log.Fatal(err)
  }
  defer func(Body io.ReadCloser) {
    err := Body.Close()
    if err != nil {
      log.Fatal(err)
    }
  }(response.Body)
  responseData, err := io.ReadAll(response.Body)
  if err != nil {
    log.Fatal(err)
  }
  if response.StatusCode != 200 {
    log.Printf("ERROR: Non-success response (%d) received from the petstore service on the query for %s.  Body received is: %s",
      response.StatusCode, opts["query"], responseData)
    return
  }

  log.Printf("Received: %s", responseData)

}

func main() {
  userOpts := parseOpts()

  zitiCtxCollection := checkCreateIdentity(userOpts)

  connectZitiService(zitiCtxCollection, userOpts)
}
